### PR TITLE
Fix default values of world id and group id to get proper collision filtering

### PR DESCRIPTION
**Pull request by nvtw**
## Description

This PR fixes inconsistencies in collision filtering logic and updates default collision group values.

### Problem

1. **Inconsistent filtering**: The `find_shape_contact_pairs` method in `builder.py` was not using the same collision filtering rules as `test_world_and_group_pair` in `broad_phase_common.py`, causing discrepancies.

2. **Invalid defaults**: Default `group_index` values were set to `-1` in the USD importer and `ShapeFlags` struct. According to the filtering rules, negative indices don't collide with themselves, so nothing was colliding. Setting default to `0` for groups also doesn't work since group `0` means "no collisions".

### Solution

1. Updated `find_shape_contact_pairs` to use the exact same filtering logic as `test_world_and_group_pair` for consistency across all collision modes.

2. Changed default collision group from `-1` to `1` in:
   - `ModelBuilder.ShapeConfig.collision_group`
   - USD importer (`parse_usd`)

3. Updated affected unit tests to reflect new default values.

### Collision Filtering Rules

The `test_world_and_group_pair` function implements these rules (agreed upon in Zurich meeting):

**Collision groups:**
- `0`: No collisions (disabled)
- Positive (`1, 2, 3, ...`): Only collide with themselves and negative groups
- Negative (`-1, -2, -3, ...`): Collide with everything except their negative counterpart

## Newton Migration Guide

Please ensure the migration guide for **warp.sim** users is up-to-date with the changes made in this PR.

- [x] The migration guide in ``docs/migration.rst`` is up-to date

## Before your PR is "Ready for review"

- [x] Necessary tests have been added and new examples are tested (see `newton/tests/test_examples.py`)
- [x] Documentation is up-to-date
- [x] Code passes formatting and linting checks with `pre-commit run -a`

## Summary by CodeRabbit

* **Bug Fixes**
  * Collision group defaults now start at 1 (0 disables collisions), fixing import and simulation behavior.

* **Refactor**
  * Broad-phase filtering aligned with kernel-style logic for more consistent and slightly faster shape-pair checks, with earlier loop exits on non-matching worlds.

* **Documentation**
  * New Collisions and Contacts guide added and linked from Concepts, clarifying worlds, groups, filtering, and performance.

* **Chores**
  * Exposed additional geometry test helpers for broader use.

### CODE REVIEW

**Comment by adenzler-nvidia** [comment]
Without looking at the code - the API does make sense to me. Will have a look at implementation once we agree the API is good.

File: newton/examples/basic/example_basic_urdf.py [hunk 1]
```diff
@@ -74,9 +74,13 @@ def __init__(self, viewer, num_worlds):
         # use "scene" for the entire set of worlds
         scene = newton.ModelBuilder()
 
+        # Add global ground plane that collides with all worlds
+        scene.current_world = -1
+        scene.add_ground_plane()
```
Comment (vreutskyy) [inline newton/examples/basic/example_basic_urdf.py]: This I don't understand. Perhaps as a C++ programmer I just don't get the spirit of Python, but still, why set this current_world variable before calling this add_* function when it obviously should be passed IN the function as an argument. And why allow using just some random number as a world ID like scene.current_world = 12341412341341? There obviously should be add_world() function incrementing num_worlds variable and returning next valid world ID. But I've already told this all. Probably there's something I don't understand why this API is so good.

Comment (nvtw) [reply newton/examples/basic/example_basic_urdf.py]: I agree, this MR is not trying to make the scene setup better in any means. All the changes in examples and tests are only a consequence of fixing bugs and inconsistencies in the collision filtering logic.
I would leave an improved scene construction API to you @vreutskyy , saw the proposed design in that google doc and to me it looks good. No idea what other people think about it.

Comment (adenzler-nvidia) [reply newton/examples/basic/example_basic_urdf.py]: I agree with Tobi, we should discuss the world API as a separate thing and use this PR to align on the collision filtering behavior/API.

Comment (vreutskyy) [reply newton/examples/basic/example_basic_urdf.py]: I'm just trying to understand the logic behind the worlds and collision groups systems, and as I'm used to game physics engines, I'm may miss some robotics oriented specifics. Like, how the collision groups are supposed do work. Often in game physics engines you have two bit masks for shape/body collision filtering: collision_groups and collides_with. The filtering works like (a.collison_groups & b.collides_with) || (b.collison_groups & a.collides_with). MuJoCo does this too - they say they took it from ODE. But we just have a single integer for collision_group. How does it work?
Also, you've changed current_world default from -1 to 0. But as for me, world0 should be explicitly created before it can be used. That's why default value is -1. But again, maybe I'm missing something.

Comment (nvtw) [reply newton/examples/basic/example_basic_urdf.py]: The problem with everything in world -1 is that Sweep and Prune cannot deal efficiently with that. World -1 is not really a world, it's a space for objects that are supposed to collide with all other worlds. So the sweep and prune actually adds a copy (an index that references that instance) of each element with worldId -1 to the sort array of each world. I have not found a good solution around this except putting everything into world 0 as the default and the user should be very careful what to put into world -1. Ideally only a single terrain mesh or very few object should have worldId -1. Please let me know if you know of a way to make the -1 world strategy work well with sweep and prune such that sorting per world is still possible (this is a significant speedup over sorting everything).
About the collision group filtering. This is confusing. The function is inherited from warp sim as far as I know and it's quite different from the ODE style filtering. In my opinion, we can also switch to ODE style filtering but others should decide that as it will affect backward compatibility.

Those are the filtering rules I was told to use:
```
# Collision filtering
@wp.func
def test_group_pair(group_a: int, group_b: int) -> bool:
    """Test if two collision groups should interact.

    Args:
        group_a: First collision group ID. Positive values indicate groups that only collide with themselves (and with negative groups).
                Negative values indicate groups that collide with everything except their negative counterpart.
                Zero indicates no collisions.
        group_b: Second collision group ID. Same meaning as group_a.

    Returns:
        bool: True if the groups should collide, False if they should not.
    """
    if group_a == 0 or group_b == 0:
        return False
    if group_a > 0:
        return group_a == group_b or group_b < 0
    if group_a < 0:
        return group_a != group_b
```

Comment (nvtw) [reply newton/examples/basic/example_basic_urdf.py]: I was thinking about it a bit more, maybe I can support all objects in worldId -1 in a better way such that we can keep the worldId -1 default. I'll give it a try. Will post again here once I know more.

Comment (nvtw) [reply newton/examples/basic/example_basic_urdf.py]: Ok, it's indeed possible to support the world id -1 case in a much better way. I reverted all changes related to it. The MR is now much smaller.

**Comment by adenzler-nvidia** [comment]
@nvtw can you please add documentation that explains the collision filtering mechanisms?

**Comment by nvtw** [comment]
@adenzler-nvidia  Where should that documentation end up? I did not find a well matching file in newton/docs, the closest would be guide/tutorials.rst. And the in-code documentation is present, see method test_group_pair and others.

**Comment by adenzler-nvidia** [comment]
Probably makes sense to start a new section next to conventions/articulations about collisions? I assume we'll have more to say about collisions soon..

should the test_group_pair be part of the public API to be included in the auto generated docs?

File: newton/_src/sim/builder.py [hunk 2]
```diff
@@ -4840,6 +4840,49 @@ def finalize(self, device: Devicelike | None = None, requires_grad: bool = False
 
             return m
 
+    def _test_group_pair(self, group_a: int, group_b: int) -> bool:
```
Comment (eric-heiden) [inline newton/_src/sim/builder.py:4843]: This function should be public and be a `@wp.func`-annotated function that we should move to `newton.geometry` I think so that it can be reused there also inside the broadphase collision kernels. I think you have this logic implemented somewhere else there already?

File: newton/_src/sim/builder.py [hunk 3]
```diff
@@ -4840,6 +4840,49 @@ def finalize(self, device: Devicelike | None = None, requires_grad: bool = False
...
+        """
+        if group_a == 0 or group_b == 0:
+            return False
+        if group_a > 0:
+            return group_a == group_b or group_b < 0
+        if group_a < 0:
+            return group_a != group_b
+        return False
+
+    def _test_world_and_group_pair(
```
Comment (eric-heiden) [inline newton/_src/sim/builder.py:4863]: Same here, I would have this as a Warp function exposed through `newton.geometry`.

**Comment by eric-heiden** [comment]
> Probably makes sense to start a new section next to conventions/articulations about collisions? I assume we'll have more to say about collisions soon..
> 
> should the test_group_pair be part of the public API to be included in the auto generated docs?

I would also vote for another conventions section on collision handling. There we can explain this world / collision group logic and also the kinds of pair-wise collision shape types we support for each collision pipeline we have: https://github.com/newton-physics/newton/issues/853 (can be a follow-up PR of course).

File: newton/_src/sim/builder.py (hunk 2 above)
Comment (eric-heiden) [reply newton/_src/sim/builder.py:4843]: One possible issue with calling a Warp function from a Python context is the big performance penalty involved currently.
If it's too severe then let's just keep it here for now.

Comment (adenzler-nvidia) [reply newton/_src/sim/builder.py:4843]: This is good to know, Is there a warp ticket to track this? I guess this would be good to have a look at.

**Comment by nvtw** [comment]
> Probably makes sense to start a new section next to conventions/articulations about collisions? I assume we'll have more to say about collisions soon..
> 
> should the test_group_pair be part of the public API to be included in the auto generated docs?

There is some documentation now and the test functions are now in the public API

File: newton/_src/sim/builder.py (hunk 3 above)
Comment (nvtw) [reply newton/_src/sim/builder.py:4863]: Done

File: newton/_src/sim/builder.py (hunk 2 above)
Comment (nvtw) [reply newton/_src/sim/builder.py:4843]: Done, I'll rely on the benchmarks to catch a possible performance regression. They already caught one when I accidentally reverted an optimization, so they do work.

Comment (nvtw) [reply newton/_src/sim/builder.py:4843]: Ok, the perf hit looks unacceptable
```
| Change   | Before [fe8a3405] <main>   | After [d212b81e]    |   Ratio | Benchmark (Parameter)                                                        |
|----------|----------------------------|---------------------|---------|------------------------------------------------------------------------------|
| +        | 25.1±0.3ms                 | 145±1ms             |    5.77 | setup.bench_model.FastInitializeModel.time_initialize_model('cartpole', 256) |
| +        | 1.17±0.01s                 | 1.32±0.01s          |    1.12 | setup.bench_model.FastInitializeModel.time_initialize_model('g1', 256)       |
| +        | 93.5±0.6ms                 | 2.25±0.01s          |   24.07 | setup.bench_model.FastInitializeModel.time_initialize_model('humanoid', 256) |
```

Comment (eric-heiden) [reply newton/_src/sim/builder.py:4843]: Thanks, let's keep the Python function then.
@christophercrouzet for viz how slow things get when we call Warp function from Python.
